### PR TITLE
21 - add project information section to application forms

### PR DESCRIPTION
**Pull request by caravinci**
This code includes the layout and validation for the input and textarea fields, but doesn't include the full validation + interaction logic for the Publication URLs, which will be added in a separate ticket

### CODE REVIEW

File: components/pages/Applications/ApplicationForm/Forms/DoubleFieldRow.tsx [hunk 1]
```diff
@@ -4,37 +4,48 @@ import { css } from '@icgc-argo/uikit/index';
 import FormFieldHelpBubble from './FormFieldHelpBubble';
 
 const DoubleFieldRow = ({
+  actions,
```
Comment (caravinci) [inline components/pages/Applications/ApplicationForm/Forms/DoubleFieldRow.tsx:7]: allows adding icons/buttons to rows

File: components/pages/Applications/ApplicationForm/Forms/DoubleFieldRow.tsx [hunk 2]
```diff
@@ -4,37 +4,48 @@ import { css } from '@icgc-argo/uikit/index';
...
+  className = '',
```
Comment (caravinci) [inline components/pages/Applications/ApplicationForm/Forms/DoubleFieldRow.tsx:8]: this was useful to customise the component in specific contexts

File: components/pages/Applications/ApplicationForm/Forms/FormFieldHelpBubble.tsx [hunk 3]
```diff
@@ -1,27 +1,35 @@
-import { ReactElement } from 'react';
-import { css, UikitTheme } from '@icgc-argo/uikit/index';
+import { ReactElement, ReactNode } from 'react';
+import { css } from '@emotion/core';
+import { UikitTheme } from '@icgc-argo/uikit/index';
 import Typography from '@icgc-argo/uikit/Typography';
 import { useTheme } from '@icgc-argo/uikit/ThemeProvider';
 
-const HelpBubble = ({ tail, text }: { tail?: 'left' | 'right'; text: string }): ReactElement => {
+const HelpBubble = ({
```
Comment (caravinci) [inline components/pages/Applications/ApplicationForm/Forms/FormFieldHelpBubble.tsx:7]: the changes here make the bubble become a banner when the user doesn't want a tail.

File: components/pages/Applications/ApplicationForm/Forms/ProjectInfo/PublicationURLs.tsx [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+import Link from '@icgc-argo/uikit/Link';
+import Typography from '@icgc-argo/uikit/Typography';
+
+import DoubleFieldRow from '../DoubleFieldRow';
+import { FormFieldType } from '../types';
+import { getMin, isRequired } from '../validations';
+
+const ID = 'publicationURLs';
+
+const PublicationURLs = ({
```
Comment (caravinci) [inline components/pages/Applications/ApplicationForm/Forms/ProjectInfo/PublicationURLs.tsx:20]: this abstracts the list of `publication URLs` and manages their state independently.
WIP

File: components/pages/Applications/ApplicationForm/Forms/ProjectInfo/index.tsx [hunk 5]
```diff
@@ -0,0 +1,306 @@
...
+import FormFieldHelpBubble from '../FormFieldHelpBubble';
+import RequiredFieldsMessage from '../RequiredFieldsMessage';
+import {
+  FormSectionValidationState_ProjectInfo,
+  FormSectionValidatorFunction_Origin,
+} from '../types';
+import { isRequired, useLocalValidation } from '../validations';
+import PublicationURLs from './PublicationURLs';
+
+const ProjectInfo = ({
```
Comment (caravinci) [inline components/pages/Applications/ApplicationForm/Forms/ProjectInfo/index.tsx:19]: the section itself
not much to see here other than the new Textarea word/char counter props

File: components/pages/Applications/ApplicationForm/Forms/constants.tsx [hunk 6]
```diff
@@ -33,7 +34,7 @@ export const sectionsOrder = [
   'applicant',
   'representative',
   'collaborators',
-  'project',
+  'projectInfo',
```
Comment (caravinci) [inline components/pages/Applications/ApplicationForm/Forms/constants.tsx:37]: makes the section available in the form

File: components/pages/Applications/ApplicationForm/Forms/index.tsx [hunk 7]
```diff
@@ -176,6 +181,19 @@ const ApplicationFormsBase = ({ appId = 'none' }): ReactElement => {
...
+                  }
+                }
+
+                [class*='DoubleFieldRow'] {
+                  margin-bottom: 10px;
+
+                  [class*='FormControl'] {
+                    margin-bottom: 0;
+                  }
                 }
```
Comment (caravinci) [inline components/pages/Applications/ApplicationForm/Forms/index.tsx:197]: styling tweaks for this section

File: components/pages/Applications/ApplicationForm/Forms/types.ts [hunk 8]
```diff
@@ -16,7 +16,9 @@ export type HonorificsListTypes = typeof honorificsList[number];
 export enum EVENT_TARGET_TAGS {
   INPUT = 'INPUT',
   MULTISELECT = 'MULTISELECT',
+  REMOVE = 'REMOVE',
   SELECT = 'SELECT',
+  TEXTAREA = 'TEXTAREA',
```
Comment (caravinci) [inline components/pages/Applications/ApplicationForm/Forms/types.ts:21]: tags for the Textarea changes and the remove field button

File: components/pages/Applications/ApplicationForm/Forms/types.ts [hunk 9]
```diff
@@ -118,7 +120,19 @@ export type FormSectionValidationState_ITAgreements =
...
+    background: { value: string };
+    methodology: { value: string };
+    publicationURLs: {
+      hasThreeValidURLs: boolean;
+      value: Record<number, FormFieldType>;
+    };
+    summary: { value: string };
+    title: { value: string };
+    website: { value: string };
+  }>;
```
Comment (caravinci) [inline components/pages/Applications/ApplicationForm/Forms/types.ts:135]: typed section fields

File: components/pages/Applications/ApplicationForm/Forms/types.ts [hunk 10]
```diff
@@ -156,7 +170,13 @@ export type FormSectionValidationState_SectionBase = {
...
 } & Partial<SchemaObjectDescription>;
 
-export type FormValidationActionTypes = 'array' | 'boolean' | 'string' | 'object' | 'overall';
+export type FormValidationActionTypes =
+  | 'array'
+  | 'boolean'
+  | 'string'
+  | 'object'
+  | 'overall'
+  | 'remove';
```
Comment (caravinci) [inline components/pages/Applications/ApplicationForm/Forms/types.ts:179]: adds this `remove` action type to the reducer

File: components/pages/Applications/ApplicationForm/Forms/validations/helpers.ts [hunk 11]
```diff
@@ -12,6 +15,23 @@ export const schemaValidator = (fieldSchema: any, value: any) =>
     error: error?.errors?.length > 0 ? error.errors : error.message,
   }));
 
+// the one same letter, in norwegian AND in swedish: i.e. highly unlikely used together
+export const LINE_JUMP_PLACEHOLDER = ' øö ';
+
+export const countWordsInString = (value: string) => {
```
Comment (caravinci) [inline components/pages/Applications/ApplicationForm/Forms/validations/helpers.ts:21]: counter validation stuff

File: components/pages/Applications/ApplicationForm/Forms/validations/helpers.ts [hunk 12]
```diff
@@ -49,9 +69,21 @@ export const getFieldDataFromEvent: FormFieldDataFromEvent = (event) => {
   return {};
 };
 
+export const getMin = (fieldData?: FormFieldType) =>
+  (fieldData?.tests?.find((test) => test.name === 'min')?.params?.min as number) || 0;
+
 export const isRequired = (fieldData?: FormFieldType) =>
   fieldData?.tests?.some((test) => test.name === 'required') || false;
 
+export const maxWords = (max: number) => ({
```
Comment (caravinci) [inline components/pages/Applications/ApplicationForm/Forms/validations/helpers.ts:78]: logic for Yup to validate max words in the Textarea

File: components/pages/Applications/ApplicationForm/Forms/validations/helpers.ts [hunk 13]
```diff
@@ -67,3 +99,20 @@ export const transformToSelectOptions = (list: Array<string | number>) =>
     content: value,
     value: value,
   }));
+
+export const uniquePublicationURLs = {
```
Comment (caravinci) [inline components/pages/Applications/ApplicationForm/Forms/validations/helpers.ts:103]: logic for Yup to validate duplicate URLs

File: components/pages/Applications/ApplicationForm/Forms/validations/index.ts [hunk 14]
```diff
@@ -33,16 +33,24 @@ export const validationReducer = (
...
+                          : [error, ...(item.error || [])],
+                      }
+                    : item,
+                )
+                .reduce(
+                  (acc, item, index) => ({
+                    ...acc,
+                    [index]: item,
+                  }),
+                  {},
```
Comment (caravinci) [inline components/pages/Applications/ApplicationForm/Forms/validations/index.ts:52]: ensure arrays are seen as an object, for ease of state management

File: components/pages/Applications/ApplicationForm/Forms/validations/index.ts [hunk 15]
```diff
@@ -92,6 +100,27 @@ export const validationReducer = (
       };
     }
 
+    case 'remove': {
```
Comment (caravinci) [inline components/pages/Applications/ApplicationForm/Forms/validations/index.ts:103]: if we want to remove a field from an array, this is how

File: components/pages/Applications/ApplicationForm/Forms/validations/index.ts [hunk 16]
```diff
@@ -133,15 +162,28 @@ export const validator: FormSectionValidatorFunction_Main =
       const [fieldName, fieldIndex, fieldOverride] = field.split('--');
       const fieldIsArray = !Number.isNaN(Number(fieldIndex));
 
+      if (fieldOverride) {
```
Comment (caravinci) [inline components/pages/Applications/ApplicationForm/Forms/validations/index.ts:165]: for actions like removing

File: components/pages/Applications/ApplicationForm/Forms/validations/index.ts [hunk 17]
```diff
@@ -133,15 +162,28 @@ export const validator: FormSectionValidatorFunction_Main =
...
         yup.reach(
           combinedSchema[origin],
           fieldIndex && fieldOverride !== 'overall' ? `${fieldName}[${fieldIndex}]` : fieldName,
         ),
         fieldOverride === 'overall'
-          ? Object.values<FormFieldType>(validationState[origin]?.fields[fieldName]?.value).map(
-              ({ value }) => value,
-            )
+          ? Object.values<FormFieldType>(validationState[origin]?.fields[fieldName]?.value)
+              .filter(({ hidden }) => !hidden)
```
Comment (caravinci) [inline components/pages/Applications/ApplicationForm/Forms/validations/index.ts:185]: only validate the shown fields

File: components/pages/Applications/ApplicationForm/Forms/validations/schemas.ts [hunk 18]
```diff
@@ -93,6 +94,19 @@ export const itAgreementsSchema = yup.object().shape({
   }),
 });
 
+export const projectInfoSchema = yup.object().shape({
```
Comment (caravinci) [inline components/pages/Applications/ApplicationForm/Forms/validations/schemas.ts:97]: new Yup schema

**Review by caravinci — COMMENTED** [review]
The usual

**Review by blabadi — APPROVED** [review]

**Pull request merged**